Fix #2622: Latex produces empty pages after title and table of contents

jfbu opened the pull request:
The 'openany' option was not correctly obeyed.

jfbu committed:
Fix #2622: Latex produces empty pages after title and table of contents

The 'openany' option was not correctly obeyed.

jfbu commented:
@tk0miya I corrected handling of abstract environment in sphinxmanual.cls, but how does Sphinx user specify abstract if not using `.. raw:: latex` ? is this legacy code from docutils latex2e writer ?

tk0miya commented on sphinx/texinputs/sphinxmanual.cls:
```diff
@@ -86,38 +87,54 @@
 % is followed by a blank page if the 'openright' option is used.
-%
+
+% Note added (Sphinx 1.4.4): the earlier code was wrong and in fact not really
```
The comments should describe behavior or purpose of macros. If you want to note the reason of changes, please note it in commit log.

tk0miya commented on sphinx/texinputs/sphinxmanual.cls:
```diff
@@ -86,38 +87,54 @@
 % used.  This eliminates a fair amount of crud in the individual document files.
-%
+
+% Note added (Sphinx 1.4.4): code was wrong in case of openany option. Further:
```
Same above.

BTW, the element after `\tableofcontents` is "Contents:" label by default of sphinx-quickstart. Are you okay to that?

jfbu replied:
I am not sure I understand, after `\tableofcontents` I see a `\phantomsection\label{index::doc}`, and indeed if I have elsewhere `:doc:`index`` this then makes a link to that destination which turns out be Chapter 1. I would have rather that destination before the table of contents. As per "Contents:" which quickstart creates in `index.rst` by default it is a nuisance in latex except if `toctree_only` is True. 

 I agree with your remarks about comments, I wll remove them. I was imagining removing them for 1.5 release. But then I need to remove obsolete legacy code comments which become inadequate now. I will make extra commit.

tk0miya commented:
I think `raw` directive is only way to write `abstract` block.

jfbu commented:
OK, I was curious how to add an abstract but I do not find documentation, and why it is treated inside sphinxmanual.cls. The problem with abstract is that report class usually reserves verso of title page for that. It has different ways of handling abstract depending on titlepage/notitlepage document option. But Sphinx code in `sphinxmanual.cls` does as if titlepage option always active.

Current situation is not good (whether openright or openany), because abstract if put by raw directive in first included file is after table of contents in latex output; and I tried with raw abstract in index.rst before toctree directive but then again it appears after table of contents.

Typically an abstract should appear on verso of title page, or on title page itself.

jfbu committed:
Fix #2622 (cont.): obey openany option

This cleans up comments from sphinxmanual.cls.

Memo: the logic behind the earlier sphinxmanual.cls is obscure, besides
the fact that it didn't work correctly if conf.py used openany option.
The parent commit corrected the implementation, but not the logic.

Were it not for the log message about added blank page, there would be
no need to patch \tableofcontents, because by default it will use
\chapter* which does the right thing. Similarly after the end of
abstract, a \chapter or \part (or \tableofcontents, once abstract is
properly handled by LaTeX writer (*)), currently one seems to be
contrived to use raw directive and the abstract is then after
\tableofcontents) will do the right thing in case of openany. Also the
\maketitle wrongly ended with \cleardoublepage, which was corrected in parent
commit, but better would be that it does nothing. Indeed it is not
logical for \maketitle, abstract or \tableofcontents to worry about what
comes next. Anything which comes next should be a \part or \chapter
which will do the right thing.

(*) currently one seems to be contrived to use raw directive and the
abstract is then after \tableofcontents) will do the right thing in case
of openany.

jfbu commented:
I explain situation in last commit log: better still would be to completely remove from sphinxmanual.cls these patches to abstract and tableofcontents, and the clear(double)page at end of \maketitle. So far I obeyed the earlier logic, simply correcting implementation so that openany option is correctly implemented.

jfbu committed:
Fix #2622 (done): correctly obey openany option

Clean sphinxmanual.cls of superfluous unneeded code. And also avoid
skipping blank pages if openany for bibliography and index.

jfbu commented:
OK, I have more radically revisited sphinxmanual.cls in latest commit. In my brief testing, this does the right thing but may need review.

The question of the abstract needs separate attention.

tk0miya commented:
Originally, sphinx was created to build the document of python. Before sphinx appeared, it was written by LaTeX. So it might he macro for abstract comes from python's.
At present, the macro might be no needed, but we can support to note abstract part in Sphinx if anyone needs it.

BTW, this works fine for me. Let's merge :-)
